### PR TITLE
Emailer agent port to modular

### PR DESCRIPTION
**Pull request by riley206-pnnl**
# Description
Ported monolithic Emailer agent into modular VOLTTRON.

## Type of change
- [x] New feature (non-breaking change which adds functionality)
- [x] This change requires a documentation update

# How Has This Been Tested?

- [x] Tested on fresh Ubuntu 22.04.2 virtual machine 

# Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] Existing unit tests pass locally with my changes
- [x] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

File: .pre-commit-config.yaml [hunk 1]
```diff
@@ -0,0 +1,30 @@
...
+  - id: check-xml
+  - id: forbid-new-submodules
+  - id: end-of-file-fixer
+  - id: trailing-whitespace
+  - id: check-merge-conflict
+  - id: no-commit-to-branch  # blocks main commits.  To bypass do git commit --allow-empty
+  - id: pretty-format-json
+
+
+#- repo: https://github.com/pre-commit/mirrors-autopep8
```
Comment (craigpnnl) [inline .pre-commit-config.yaml]: Go ahead and remove this... commented out stuff and add google's yapf
```
# For more information about YAPF, see https://github.com/google/yapf
- repo: https://github.com/google/yapf
  rev: v0.40.2  # Use the sha / tag you want to point at
  hooks:
  - id: yapf
    name: yapf
    description: "A formatter for Python files."
    entry: yapf
    language: python
    types: [ python ]
```

File: README.md [hunk 2]
```diff
@@ -1 +1,200 @@
-# volttron-emailer
\ No newline at end of file
+volttron-emailer
```
Comment (craigpnnl) [inline README.md]: Why did you change the header and add the line to it?

Start the first paragraph with the `volttron-emailer` allows any agent on a VOLTTRON instance to send an email.  Then you can link to the alert agent and continue on .

File: README.md [hunk 3]
```diff
@@ -1 +1,200 @@
...
+
+Agents needing to send an email through the instance can do so by
+sending the following header and message to the `platform/send_email` topic
+which is monitored by the Emailer agent.  The following 
+is the expected payload for the message body and the optional header.
+
+
+# Prerequisites
+
+* Python 3.8
```
Comment (craigpnnl) [inline README.md]: So this requires python 3.10 as volttron requires 3.10 for modular.

Please modify

File: README.md [hunk 4]
```diff
@@ -1 +1,200 @@
...
+2. Install volttron and start the platform.
+
+    ```shell
+    pip install volttron
+
+    # Start platform with output going to volttron.log
+    volttron -vv -l volttron.log &
+    ```
+
+3. Create you configuration file, in this example the file is named `config`
```
Comment (craigpnnl) [inline README.md]: How do you configure tls?  Do you still have all the other items?  Can we create a configuration overview file.

File: README.md [hunk 5]
```diff
@@ -1 +1,200 @@
...
+    ```
+4. Install and start emailer agent.
+
+    ```shell
+    vctl install volttron-emailer --tag email --agent-config <path to config> --start --force
+    ```
+
+## Testing
+
+To verfiy functionality, Get the [send_email.py](./tests/send_email.py) file and run it with
```
Comment (craigpnnl) [inline README.md]: Perhaps we should tell the user what to expect as output from this and what that script does?  What about a failure what happens?

File: README.md [hunk 6]
```diff
@@ -1 +1,200 @@
...
+
+```bash
+python3 send_email.py
+```
+If all goes well, you should see an email from the address you specified in the config file. 
+
+Optional Headers
+----------------
+
+Emails by default will be sent to the initial configured email
```
Comment (craigpnnl) [inline README.md]: What's a header?  How does the user set a header?

File: config [hunk 7]
```diff
@@ -0,0 +1,15 @@
+{
```
Comment (craigpnnl) [inline config]: rename this file to example.config.json and use it as the example.

Comment (craigpnnl) [reply config]: OR better change it to yaml format and have comments everywhere.

File: README.md [hunk 8]
```diff
@@ -1 +1,200 @@
...
+
+    # When an alert is sent typically it can have the effect of being
+    # sent many times.  This setting throttles the sending of email only
+    # after a specific number of minutes.
+    #
+    # DEFAULT: "allow-frequency-minutes": 60
+    "allow-frequency-minutes": 120
+}
+```
+# Disclaimer Notice
```
Comment (craigpnnl) [inline README.md]: Not a correct level for md

File: pyproject.toml [hunk 9]
```diff
@@ -0,0 +1,62 @@
+[build-system]
+requires = ["poetry-core>=1.2.2"]
+build-backend = "poetry.core.masonry.api"
+
+[tool.poetry.scripts]
+volttron-emailer = "emailer.agent:main"
+
+[tool.isort]
```
Comment (craigpnnl) [inline pyproject.toml:8]: we should add isort to pre-commit

File: src/emailer/agent.py [hunk 10]
```diff
@@ -0,0 +1,380 @@
...
+from volttron.client.vip.agent import Agent, Core, PubSub
+from volttron import utils
+from volttron.client.messaging import topics
+from volttron.client.messaging.health import ALERT_KEY, STATUS_BAD, Status, \
+    STATUS_GOOD
+from volttron.client.vip.agent import Agent
+
+utils.setup_logging()
+_log = logging.getLogger(__name__)
+__version__ = '1.4.2'
```
Comment (craigpnnl) [inline src/emailer/agent.py]: This version should not be different than the version installed...we need to sync these better...

**Review by craigpnnl — APPROVED** [review]
Please consult a markdown document linter for vs code, it will tell you there is a bunch of issues in the markdown file you have here.

**Review by craigpnnl — CHANGES_REQUESTED** [review]
I pressed the wrong button :(

**Comment by craigpnnl** [comment]
Tests MUST start with test_ in order for them to be run automatically.  Please add a test that will send using the TestServer on volttron-testing.

File: README.md (hunk 2 above)
Comment (riley206-pnnl) [reply README.md]: I put volttron-emailer there because the other agents had that, I just forgot to add the #, also I could not find the modular alert agent to link to.

File: src/emailer/agent.py (hunk 10 above)
Comment (riley206-pnnl) [reply src/emailer/agent.py]: with the version installed. What version of what? the volttron version or the agent watcher volttron? what is this for?

Comment (craigpnnl) [reply src/emailer/agent.py]: this is the version of this particular agent
i.e. the pip version
